src/dft/backends/cufft/commit.cpp

```diff
@@ -0,0 +1,231 @@
+        std::copy(config_values.dimensions.begin(), config_values.dimensions.end(), n_copy.data());
+        const int rank = static_cast<int>(config_values.dimensions.size());
+        const int istride = static_cast<int>(config_values.input_strides.back());
+        const int ostride = static_cast<int>(config_values.output_strides.back());
+        const int batch = static_cast<int>(config_values.number_of_transforms);
+        const int fwd_dist = static_cast<int>(config_values.fwd_dist);
+        const int bwd_dist = static_cast<int>(config_values.bwd_dist);
+        std::array<int, max_supported_dims> inembed;
+        if (rank == 2) {
+            inembed[1] = config_values.input_strides[1];
+        }
+        else if (rank == 3) {
+            inembed[2] = config_values.input_strides[2];
+            inembed[1] = config_values.input_strides[1] / inembed[2];
+        }
+        std::array<int, max_supported_dims> onembed;
+        if (rank == 2) {
+            onembed[1] = config_values.output_strides[1];
+        }
+        else if (rank == 3) {
+            onembed[2] = config_values.output_strides[2];
+            onembed[1] = config_values.output_strides[1] / onembed[2];
+        }
+
+        cufftHandle fwd_plan, bwd_plan;
+
+        // forward plan
+        auto res = cufftPlanMany(&fwd_plan, // plan
+                                 rank, // rank
+                                 n_copy.data(), // n
+                                 inembed.data(), // inembed
+                                 istride, // istride
+                                 fwd_dist, // idist
+                                 onembed.data(), // onembed
+                                 ostride, // ostride
+                                 bwd_dist, // odist
+                                 fwd_type, // type
+                                 batch // batch
+        );
+
+        if (res != CUFFT_SUCCESS) {
+            throw mkl::exception("dft/backends/cufft", __FUNCTION__,
+                                 "Failed to create forward cuFFT plan.");
+        }
+
+        // flip fwd_distance and bwd_distance because cuFFt uses input distance and output distance.
+        // backward plan
+        res = cufftPlanMany(&bwd_plan, // plan
+                            rank, // rank
+                            n_copy.data(), // n
+                            inembed.data(), // inembed
+                            istride, // istride
+                            bwd_dist, // idist
+                            onembed.data(), // onembed
+                            ostride, // ostride
+                            fwd_dist, // odist
+                            bwd_type, // type
+                            batch // batch
+        );
+        if (res != CUFFT_SUCCESS) {
+            throw mkl::exception("dft/backends/cufft", __FUNCTION__,
+                                 "Failed to create backward cuFFT plan.");
+        }
+        plans = { fwd_plan, bwd_plan };
```

Comment:
I believe you will have the same problems than for the CPU backend.
For real FFT there should be some cases where the input and output strides will be "valid" strides only for one of the descriptor (forward or backward descriptor) and here since you are forcing both plans to be created successfully then you might always fail to commit for some valid real FFT case.

Reply:
Thank you for pointing this out. cuFFT has ignored this so far, but I suppose a future update may change that and it would be awkward if it broke oneMKL! I took a bit of a different approach and checked the validity of the strides to decide if the plan should be created. This is done in https://github.com/oneapi-src/oneMKL/pull/284/commits/e34aae2d6f0ddaa07a8ac1fd18877b8d0c5b0782.